llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp

```diff
@@ -910,6 +916,67 @@ Instruction *InstCombinerImpl::visitTrunc(TruncInst &Trunc) {
     Changed = true;
   }
 
+  // Handle scenarios like trunc i64 (sub(ptr2int, ptr2int)) to i32 / trunc i64
```

Comment:
Shouldn't this logic be inside computeKnownBits() for sub? The information we get out of this is that the top bits of the sub result are zero. That will result in the nuw flag on trunc being set down the line.